### PR TITLE
Fixes in preparation of new build system

### PR DESCRIPTION
**Pull request by mstormi**
Fixes: #639 

Signed-off-by: Markus Storm markus.storm@gmx.net

### CODE REVIEW

File: build-image/first-boot.bash [hunk 1]
```diff
@@ -134,12 +134,11 @@ sleep 10  # Related to: https://github.com/openhab/openhabian/issues/441#issueco
 echo "OK"
 
 echo -n "$(timestamp) [openHABian] Updating repositories and upgrading installed packages... "
-apt-get update &>/dev/null
```
Comment (EliasGabrielsson) [inline build-image/first-boot.bash]: Why is this removed? 
Is essential to install new packages based on updated apt repositories.

File: functions/helpers.bash [hunk 2]
```diff
@@ -132,3 +132,18 @@ is_xenial() {
 running_in_docker() {
   grep -q 'docker\|lxc' /proc/1/cgroup
 }
+tryUntil() {
```
Comment (EliasGabrielsson) [inline functions/helpers.bash]: I would like this function to have some unit test so we know that it is working properly.

File: functions/java-jre.bash [hunk 3]
```diff
@@ -18,7 +18,7 @@ java_webupd8_archive() {
 }
 
 java_zulu(){
-  cond_redirect systemctl stop openhab2.service
+  cond_redirect systemctl is-active --quiet openhab2 && systemctl stop openhab2
```
Comment (EliasGabrielsson) [inline functions/java-jre.bash]: Nice clean solution! 👍

File: build-image/first-boot.bash (hunk 1 above)
Comment (mstormi) [reply build-image/first-boot.bash]: it's redundant and slows down the install process
apt update happens early on unattended install and/or whenever user select one of the upgrade options from the menu anyway

File: functions/helpers.bash (hunk 2 above)
Comment (mstormi) [reply functions/helpers.bash]: I've created tests in functions/helpers.bats that work from a shell but fail when used with BATS .. it is driving me crazy, please have yourrrself a look.

File: functions/java-jre.bash (hunk 3 above)
Comment (BClark09) [reply functions/java-jre.bash]: AFAIK, `systemctl stop` doesn't run the stop executable if the service isn't active so would this be necessary?

Comment (mstormi) [reply functions/java-jre.bash]: I think you're right from the logics perspective but it spits a warning which is what I want to avoid for sake of a clean output.
Ok, re-added the .service but just because I'm too lazy to extinct it everywhere else.

File: build-image/first-boot.bash (hunk 1 above)
Comment (EliasGabrielsson) [reply build-image/first-boot.bash]: Oh I thought this was the first instance of `apt-get update`. NVM.
I actually would like to remove all `apt-get update` beside:
a) Run in first-boot
b) Run when openhabian-config start
c) Run when a new repository is added.

I will fix that.

File: functions/helpers.bats [hunk 4]
```diff
@@ -2,6 +2,24 @@
 
 load helpers
 
+testthis() {
```
Comment (EliasGabrielsson) [inline functions/helpers.bats]: This test is not static in the sense that it need external dependencies.
I think we can do better, lets think about it.

[ ] We need a test which works on a local system 
[ ] Make sure Bash get exit code are forwarded somehow

File: functions/helpers.bats [hunk 5]
```diff
@@ -2,6 +2,24 @@
...
+
+@test "unit-tryExistingSite" {
+  testthis "http://www.openhab.org/" 5
+  [ "$status" -eq 0 ]
+  echo -e "# \n\e[32mGetting www.openhab.org in tryUntil loop succeeded." >&3
+}
+
+@test "unit-tryNonExistingSite" {
+  run testthis "http://nothisdoesnotexistforsure/" 0
+  [ "$status" -eq 0 ]
```
Comment (EliasGabrielsson) [inline functions/helpers.bats]: Its strange to have status "0" during a failure. 
We need to think twice about this. I will give it some effort in the end of the week.

File: functions/helpers.bash (hunk 2 above)
Comment (EliasGabrielsson) [reply functions/helpers.bash]: I will look into this in the end of the week. Don't get me wrong I really want this function in the codebase but we need to have firm tests to first. I don't want to introduce more technical debt while trying to iron out old code.

File: functions/helpers.bats (hunk 4 above)
Comment (mstormi) [reply functions/helpers.bats]: Ok, here's a better version. Duh, this cost me nerves so please accept it like that.

File: functions/helpers.bats (hunk 5 above)
Comment (mstormi) [reply functions/helpers.bats]: It's now the number-of-attempts-left which is what tryUntil returns.

File: functions/helpers.bash (hunk 2 above)
Comment (mstormi) [reply functions/helpers.bash]: 'cost me quite some nerves but here you are now.

**Comment by mstormi** [comment]
Mmh, Github still says you've requested changes but I don't see what that is targeting. I've added a unit test for tryUntil. If there's anything left please raise this again
(and yes I remember you want to do that shellcheck one first)

**Comment by mstormi** [comment]
note to myself: there's another occurence of "try ping 9.9.9.9 until ..." in first-boot.bash

**Comment by mstormi** [comment]
@EliasGabrielsson You mentioned it's just a little fix you would need to apply to get the builds to succeed again, but are you aware that the current build system actually clones and then build-tests *current master* but not the repo to contain the PR ?

That one almost drove me mad when developing for the new build system.
Hence my fix in https://github.com/mstormi/openhabian/blob/buildfixes/build-image/first-boot.bash#L146

File: build-image/first-boot.bash [hunk 6]
```diff
@@ -27,7 +31,7 @@ source /etc/openhabian.conf
 echo "OK"
 
 echo -n "$(timestamp) [openHABian] Starting webserver with installation log... "
-if hash python 2>/dev/null; then
+if hash python3 2>/dev/null; then
```
Comment (holgerfriedrich) [inline build-image/first-boot.bash:34]: python3 again... what is the reason to change.
for my current production box, default still points to 2.7, though both 2 and 3 are installed.

File: build-image/webif.bash [hunk 7]
```diff
@@ -11,7 +11,7 @@ port=80 # Port the webserver is listing to
 source /etc/openhabian.conf # to get the hostname
 
 if [ "$1" = "start" ]; then
-  mkdir /tmp/webif
+  mkdir /tmp/webif 2>/dev/null
```
Comment (holgerfriedrich) [inline build-image/webif.bash:14]: maybe  `mkdir -p /tmp/webif` is the better choice, not complaining about existing directories, and reporting severe errors

File: build-image/webif.bash [hunk 8]
```diff
@@ -39,6 +39,8 @@ if [ "$1" = "reinsure_running" ]; then
 fi
 
 if [ "$1" = "inst_done" ]; then
+  # dir was eventually deleted by reboot 
+  mkdir /tmp/webif 2>/dev/null
```
Comment (holgerfriedrich) [inline build-image/webif.bash:43]: see above

File: functions/helpers.bash [hunk 9]
```diff
@@ -148,3 +148,18 @@ is_bionic() {
...
```
Comment (holgerfriedrich) [inline functions/helpers.bash:151]: `#tryuntil <paramters> does ...`

File: openhabian-setup.sh [hunk 10]
```diff
@@ -59,7 +59,7 @@ OLDWD=$(pwd) && cd /opt
 if [[ -n "$UNATTENDED" ]]; then
   # apt/dpkg commands will not try interactive dialogs
   export DEBIAN_FRONTEND=noninteractive
-  apt-get update
+  apt-get update >/dev/null
```
Comment (holgerfriedrich) [inline openhabian-setup.sh:62]: hm, does a cond_... make sense here?

Comment (mstormi) [reply openhabian-setup.sh]: yes but I simply undid this here because it is to change anyway in #724

File: build-image/webif.bash (hunk 7 above)
Comment (mstormi) [reply build-image/webif.bash]: ok

File: functions/helpers.bash (hunk 9 above)
Comment (mstormi) [reply functions/helpers.bash:151]: ok, added

File: build-image/webif.bash (hunk 8 above)
Comment (mstormi) [reply build-image/webif.bash]: ok

File: .travis.yml [hunk 11]
```diff
@@ -1,25 +1,47 @@
 language: bash
 
+dist: disco
```
[the remark this answers is not in the document]
Comment (mstormi) [reply .travis.yml:3]: You got it right but frankly I don't fully remember. IIRC it was needed to get the latest version on one of the tools we use (shellsheck ?).
I think I can reset to bionic but OTOH it works and it's in line with the latest'n greatest strategy.

File: .travis.yml [hunk 12]
```diff
@@ -1,25 +1,47 @@
...
+        - docker commit "openhabian-test" openhabian-installed
+    - stage: Bats
+      script:
+        docker run --name "bats-inst-tests" --privileged -d openhabian/rpi-openhabian
+        docker exec -it "bats-inst-tests" bash -c 'bats -r -f "installation-." .'
+    - script:
+        docker run --name "bats-dest-tests" --privileged -d openhabian/rpi-openhabian
+        docker exec -it "bats-dest-tests" bash -c 'bats -r -f "destructive-." .'
+    - stage: Shellcheck
+      script: shellcheck -x -s bash openhabian-setup.sh || true
```
[the remark this answers is not in the document]
Comment (mstormi) [reply .travis.yml:45]: ok

File: .travis.yml [hunk 13]
```diff
@@ -1,25 +1,47 @@
...
+jobs:
+  include:
+    - stage: Dev+unit-tests
+      script: docker run -it openhabian/rpi-openhabian bash -c 'bats -r -f "dev-." .'
+    - script: docker run -it openhabian/rpi-openhabian bash -c 'sudo bats -r -f "unit-." .'
+    - stage: Openhabian-Setup
+      script:
+        - docker run --name "openhabian-test" --privileged -d openhabian/rpi-openhabian
+        - docker exec -it "openhabian-test" bash -c \
+            'VIRT=$(sudo virt-what); echo -e "\n\e[34mTesting openhabian install on ${VIRT} ...\n"; ./build.bash local-test && mv ~/.profile ~/.bash_profile && /etc/rc.local; OUTPUT=$?; if [ "${VIRT}" != "native HW" ]; then echo -e "\n\e[32mDid not test openhabian install on native HW, ignoring test result.\n"; exit 0; fi'
```
[the remark this answers is not in the document]
Comment (mstormi) [reply .travis.yml:35]: Elias wanted to have BATS unit tests for everything new and zram.bats requires ZRAM deployment. IIRC that was failing in emulation mode so I detect emulation and skip tests.

File: .travis.yml [hunk 14]
```diff
@@ -1,25 +1,47 @@
...
-  - docker build --tag openhabian/openhabian-bats .
-
-script:
-  - docker run -it openhabian/openhabian-bats bash -c 'bats -r -f "unit-." .'
-  - docker run --name "install-test" --privileged -d openhabian/openhabian-bats
-  - travis_wait 60 sleep infinity & docker exec -it "install-test" bash -c "./build.bash local-test && mv ~/.profile ~/.bash_profil && /etc/rc.local"
-  - travis_wait 60 sleep infinity & docker exec -it install-test bash -c 'bats -r -f "installation-." .'
-  - travis_wait 60 sleep infinity & docker exec -it install-test bash -c 'bats -r -f "destructive-." .'
+  - sudo apt-get install virt-what && VIRT=$(sudo virt-what) && echo "Travis VM running on ${VIRT:-native HW}."
+  - docker run --rm --privileged multiarch/qemu-user-static:register --reset
```
[the remark this answers is not in the document]
Comment (mstormi) [reply .travis.yml:24]: added comment

File: .travis.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (holgerfriedrich) [reply .travis.yml:36]: For me it was not clear, _which_ tests were skipped from reading the code here....

[the remark this answers is not in the document]
Comment (mstormi) [reply .travis.yml:36]: it's ZRAM only for now see functions/zram.bats

File: .travis.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (holgerfriedrich) [reply .travis.yml:3]: fine for me, it lifts shellchek to the same version as in Debian Buster
https://launchpad.net/ubuntu/+source/shellcheck

File: build-image/first-boot.bash (hunk 6 above)
Comment (mstormi) [reply build-image/first-boot.bash:34]: The webserver for the webif to show install status (boot.log) didn't work in my testing.
Now if only I could remember all the details 9 months after ... but the fix was to move to Python3. As I didn't want to upgrade to P3 I just installed and explicitly called it.

Comment (holgerfriedrich) [reply build-image/first-boot.bash:34]: should not harm, but your explanation about webif problem with python2 is worth a comment in the code....
For mr ok to keep as py27 will be removed at some point in time anyway...

Comment (mstormi) [reply build-image/first-boot.bash:34]: ok, added a comment

So we're done ?
Merge ? I'll be at home tomorrow 😅

File: Dockerfile [hunk 15]
```diff
@@ -1,6 +1,8 @@
-FROM minimum2scp/systemd-stretch
+FROM minimum2scp/systemd-buster
+#FROM minimum2scp/systemd-stretch
+#FROM balenalib/intel-nuc-debian-node:latest-buster
 
-RUN apt-get update && apt-get install -y git locales jq
+RUN apt-get update && apt-get install -q -y git locales python3 python3-pip jq virt-what
```
[the remark this answers is not in the document]
Comment (mstormi) [reply Dockerfile:5]: for webif, see below

File: .travis.yml [hunk 16]
```diff
@@ -1,25 +1,50 @@
...
+jobs:
+  include:
+    - stage: Dev+unit-tests
+      script: docker run -it openhabian/rpi-openhabian bash -c 'bats -r -f "dev-." .'
+    - script: docker run -it openhabian/rpi-openhabian bash -c 'sudo bats -r -f "unit-." .'
+    - stage: Openhabian-Setup
+      script:
+        - docker run --name "openhabian-test" --privileged -d openhabian/rpi-openhabian
+        - docker exec -it "openhabian-test" bash -c \
+            'VIRT=$(sudo virt-what); echo -e "\n\e[34mTesting openhabian install on ${VIRT} ...\n"; ./build.bash local-test && mv ~/.profile ~/.bash_profile && /etc/rc.local; OUTPUT=$?; if [ "${VIRT}" != "native HW" ]; then echo -e "\n\e[32mDid not test openhabian install on native HW, ignoring test result.\n"; exit 0; fi'
```
Comment (holgerfriedrich) [inline .travis.yml:37]: Do you think it is a good idea to ignore _all_ errors from local-test run?
For now, it might be the way to go, but IMHO we should resolve this within the test code where required or split up the local tests....

File: build-image/webif.bash [hunk 17]
```diff
@@ -31,6 +31,7 @@ if [ "$1" = "start" ]; then
   (cd /tmp/webif || exit 1; python3 -m http.server $port > /dev/null 2>&1 &)
 fi
 
+# changed to python3 due to issues in automated testing
```
Comment (holgerfriedrich) [inline build-image/webif.bash:34]: ...what you explained was not related to automated testing, it was about the webif, right?

Comment (mstormi) [reply build-image/webif.bash:34]: yes it was the webif to fail, but the occasion this showed up was during automated testing

File: .travis.yml (hunk 16 above)
Comment (mstormi) [reply .travis.yml:37]: No of course we should not ignore any tests but I couldn't think of a proper split at this stage.
Let's review this when we got the CI working.

Comment (mstormi) [reply .travis.yml:37]: Created #831

**Comment by mstormi** [comment]
ok, here we go

**Comment by mstormi** [comment]
He-ho:
https://travis-ci.com/github/openhab/openhabian/builds/158544684?utm_medium=notification&utm_source=email

**Comment by mstormi** [comment]
@holgerfriedrich Travis processes the "install" section but does not proceed to process "jobs"
Any idea what's missing?
